x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/IndexRequestCreator.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.persistence;
+
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.core.ml.utils.ToXContentParams;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * A helper class for creating index requests that handles determining whether the model being indexed can be overwritten if it already
+ * exists.
+ */
+public class IndexRequestCreator {
+    private static final ToXContent.Params FOR_INTERNAL_STORAGE_PARAMS = new ToXContent.MapParams(
+        Collections.singletonMap(ToXContentParams.FOR_INTERNAL_STORAGE, "true")
+    );
+
+    /**
+     * This is package private so the tests can access it
+     */
+    static final Set<String> PRE_PACKAGED_MODELS = Set.of(".elser_model_1");
```

Comment:
This strings `.elser_model_1` also exists at: https://github.com/jonathan-buttner/elasticsearch/blob/ml-put-trained-allow-overwrite/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartTrainedModelDeploymentAction.java#L658

Any thoughts on a good central place to put this?

Reply:
hard coding in code has a big disadvantage: We will never be able to push out a new model after a release is out. It is also extra burden to maintain this list and keep it in sync. I think we don't want to go this way.

The `isElserModel` method has been implemented as a short term solution, because we struggled with estimating the required memory usage. It turned out it is different to the estimation we have for other pytorch models, but we wanted to play safe, that's why we implemented it this way.

Long - hopefully not so long - term the plan is to replace the estimate and let the model define its memory usage as part of the metadata.

Long story short: I think the decision here should not be based on string matching, but simply whether it is a pre-packaged model or not.

Reply:
I suggest to add a parameter to `PutTrainedModelDefinitionPartAction.Request`, e.g. `overwrite` with a default of `false`. `ModelImporter` sets this to `true`. If `true` you use the index not the create option.

Note: There is no immediate need to expose `overwrite` in the REST layer, we can keep this internal.

Reply:
Thanks for the background and that makes sense. Just to make sure I understand, you're suggesting instead of having the the `IndexRequestCreator` have a `Set` to determine whether it is a prepackaged model, we have the `ModelImporter` send a flag indicating whether we should overwrite and pass that down through to the `TrainedModelProvider` which will determine whether we use `Index` or `Create`?

Reply:
yes, if it originates from `ModelImporter` we know it originates from a pre-packaged model.